src/main/java/fi/hsl/transitdata/tripupdate/gtfsrt/GtfsRtFactory.java

```diff
@@ -18,63 +18,75 @@
         } else {
+            String stopId = stopEstimate.getStopId();
+            int stopSequence = stopEstimate.getStopSequence();
             stopTimeUpdateBuilder = GtfsRealtime.TripUpdate.StopTimeUpdate.newBuilder()
-                    .setStopId(String.valueOf(stopEvent.getRouteData().getStopId()))
-                    .setStopSequence(stopEvent.getStopSeq());
+                    .setStopId(stopId)
+                    .setStopSequence(stopSequence);
         }
 
-        switch (stopEvent.getScheduleRelationship()) {
-            case Skipped:
+        switch (stopEstimate.getStatus()) {
+            case SKIPPED:
                 stopTimeUpdateBuilder.setScheduleRelationship(GtfsRealtime.TripUpdate.StopTimeUpdate.ScheduleRelationship.SKIPPED);
                 break;
-            case Scheduled:
+            case SCHEDULED:
                 stopTimeUpdateBuilder.setScheduleRelationship(GtfsRealtime.TripUpdate.StopTimeUpdate.ScheduleRelationship.SCHEDULED);
                 break;
         }
+        // GTFS-RT treats times in seconds
+        long stopEventTimeInSeconds = stopEstimate.getEstimatedTimeUtcMs() / 1000;
 
         GtfsRealtime.TripUpdate.StopTimeEvent stopTimeEvent = GtfsRealtime.TripUpdate.StopTimeEvent.newBuilder()
-                .setTime(stopEvent.getTargetTime())
+                .setTime(stopEventTimeInSeconds)
                 .build();
-        switch (stopEvent.getEventType()) {
-            case Arrival:
+        switch (stopEstimate.getType()) {
+            case ARRIVAL:
                 stopTimeUpdateBuilder.setArrival(stopTimeEvent);
                 break;
-            case Departure:
+            case DEPARTURE:
                 stopTimeUpdateBuilder.setDeparture(stopTimeEvent);
                 break;
         }
 
         return stopTimeUpdateBuilder.build();
     }
 
-    public static GtfsRealtime.TripUpdate newTripUpdate(StopEvent event) {
+    public static long lastModified(InternalMessages.StopEstimate estimate) {
+        return estimate.getLastModifiedUtcMs() / 1000;
+    }
+
+    public static GtfsRealtime.TripUpdate newTripUpdate(InternalMessages.StopEstimate estimate) {
+        final String routeName = reformatRouteName(estimate.getTripInfo().getRouteId());
+        final int direction = PubtransFactory.joreDirectionToGtfsDirection(estimate.getTripInfo().getDirectionId());
 
         GtfsRealtime.TripDescriptor tripDescriptor = GtfsRealtime.TripDescriptor.newBuilder()
-                .setRouteId(reformatRouteName(event.getRouteData().getRouteName()))
-                .setDirectionId(event.getRouteData().getDirection())
-                .setStartDate(event.getRouteData().getOperatingDay())
-                .setStartTime(event.getRouteData().getStartTime())
+                .setRouteId(routeName)
+                .setDirectionId(direction)
+                .setStartDate(estimate.getTripInfo().getOperatingDay()) // Local date as String
```

Comment:
I wonder what the date is here after midnight?

Reply:
at least according to [this](https://developers.google.com/transit/gtfs/reference/#frequenciestxt) also gtfs supports times > 24:00. I assume the operating day is then the previous date. These come directly from pubtrans in our case